### PR TITLE
Add NormOperator and DistOperator

### PR DESCRIPTION
**Pull request by adler-j**
Minor pull, noticed that these guys were missing.

### CODE REVIEW

File: odl/operator/default_ops.py [hunk 1]
```diff
@@ -579,6 +585,211 @@ def __str__(self):
...
+        >>> derivative = op.derivative([1, 0, 0])
+        >>> derivative([1, 0, 0])
+        1.0
+        """
+        point = self.domain.element(point)
+        norm = point.norm()
+        if norm == 0:
+            raise ValueError('not differentiable in 0')
+
+        return InnerProductOperator(point / norm)
```
Comment (kohr-h) [inline odl/operator/default_ops.py]: This assumes quite a bit on the norm of the space. The inner product operator needs to use the same weighting as the norm - I guess that's okay. But a custom norm implementation cannot be handled consistently, that needs to be filtered out (and documented that it won't work).

File: odl/operator/default_ops.py [hunk 2]
```diff
@@ -579,6 +585,211 @@ def __str__(self):
...
+            defined in the Frechet sense.
+
+        Notes
+        -----
+        The derivative cannot be written in a general sense except in Hilbert
+        spaces, in that case, it is given by
+
+        .. Math::
+
+            (D ||.||)(x)(y) = < x / ||x||, y >
```
Comment (kohr-h) [inline odl/operator/default_ops.py]: Notation for the differential is `D`?

File: odl/operator/default_ops.py [hunk 3]
```diff
@@ -579,6 +585,211 @@ def __str__(self):
...
+        ValueError
+            If ``point.norm() == 0``, in which case the derivative is not well
+            defined in the Frechet sense.
+
+        Notes
+        -----
+        The derivative cannot be written in a general sense except in Hilbert
+        spaces, in that case, it is given by
+
+        .. Math::
```
Comment (kohr-h) [inline odl/operator/default_ops.py]: Will that work capitalized?

File: odl/operator/default_ops.py [hunk 4]
```diff
@@ -579,6 +585,211 @@ def __str__(self):
...
+        >>> import odl
+        >>> r2 = odl.rn(2)
+        >>> op = NormOperator(r2)
+        >>> op([3, 4])
+        5.0
+        """
+        return x.norm()
+
+    def derivative(self, point):
+        """The derivative operator.
```
Comment (kohr-h) [inline odl/operator/default_ops.py]: Derivative of this operator in ``point``.

File: odl/operator/default_ops.py [hunk 5]
```diff
@@ -579,6 +585,211 @@ def __str__(self):
...
+        Raises
+        ------
+        ValueError
+            If ``point.norm() == 0``, in which case the derivative is not well
+            defined in the Frechet sense.
+
+        Notes
+        -----
+        The derivative cannot be written in a general sense except in Hilbert
+        spaces, in that case, it is given by
```
Comment (kohr-h) [inline odl/operator/default_ops.py]: , in which case it is given by

File: odl/operator/default_ops.py [hunk 6]
```diff
@@ -579,6 +585,211 @@ def __str__(self):
         return "{}.T".format(self.vector)
 
 
+class NormOperator(Operator):
```
Comment (kohr-h) [inline odl/operator/default_ops.py]: blank line

File: odl/operator/default_ops.py [hunk 7]
```diff
@@ -579,6 +585,211 @@ def __str__(self):
...
+    This is only applicable in normed spaces.
+
+    See Also
+    --------
+    InnerProductOperator : Inner product with fixed vector.
+    DistOperator : Distance to fixed vector.
+    """
+
+    def __init__(self, space):
+        """Initialize a NormOperator instance.
```
Comment (kohr-h) [inline odl/operator/default_ops.py]: I wouldn't hard-code the class name here, no need for that.

File: odl/operator/default_ops.py [hunk 8]
```diff
@@ -579,6 +585,211 @@ def __str__(self):
...
+        """
+        super().__init__(space, RealNumbers(), linear=False)
+
+    def _call(self, x):
+        """Return the norm of x.
+
+        Parameters
+        ----------
+        x : `domain` `element`
+            A vector in the domain.
```
Comment (kohr-h) [inline odl/operator/default_ops.py]: The vector thing again :-)

File: odl/operator/default_ops.py [hunk 9]
```diff
@@ -579,6 +585,211 @@ def __str__(self):
...
+
+        Returns
+        -------
+        out : ``field`` `element`
+            Result of the inner product calculation
```
Comment (kohr-h) [inline odl/operator/default_ops.py]: inner product?
dot

File: odl/operator/default_ops.py [hunk 10]
```diff
@@ -579,6 +585,211 @@ def __str__(self):
...
+        point = self.domain.element(point)
+        norm = point.norm()
+        if norm == 0:
+            raise ValueError('not differentiable in 0')
+
+        return InnerProductOperator(point / norm)
+
+    def __repr__(self):
+        """Return ``repr(self)``."""
+        return 'NormOperator({!r})'.format(self.domain)
```
Comment (kohr-h) [inline odl/operator/default_ops.py]: Use `self.__class__.__name__`

File: odl/operator/default_ops.py [hunk 11]
```diff
@@ -579,6 +585,211 @@ def __str__(self):
...
+
+    def __str__(self):
+        """Return ``str(self)``."""
+        return 'NormOperator({})'.format(self.domain)
```
Comment (kohr-h) [inline odl/operator/default_ops.py]: Use `self.__class__.__name__`

File: odl/operator/default_ops.py [hunk 12]
```diff
@@ -579,6 +585,211 @@ def __str__(self):
...
+
+
+class DistOperator(Operator):
```
Comment (kohr-h) [inline odl/operator/default_ops.py]: blank row

File: odl/operator/default_ops.py [hunk 13]
```diff
@@ -579,6 +585,211 @@ def __str__(self):
...
+    This is only applicable in metric spaces.
+
+    See Also
+    --------
+    InnerProductOperator : Inner product with fixed vector.
+    NormOperator : Norm of a vector.
+    """
+
+    def __init__(self, vector):
+        """Initialize a DistOperator instance.
```
Comment (kohr-h) [inline odl/operator/default_ops.py]: see above

File: odl/operator/default_ops.py [hunk 14]
```diff
@@ -579,6 +585,211 @@ def __str__(self):
...
+
+        Parameters
+        ----------
+        space : `LinearSpace`
+            Space to take the norm in.
+        """
+        super().__init__(space, RealNumbers(), linear=False)
+
+    def _call(self, x):
+        """Return the norm of x.
```
Comment (kohr-h) [inline odl/operator/default_ops.py]:
```
``x``
```
We have a bit mixed style for the `_call` docstring. Sometimes we say
```
Return ``self(x)``.
```
or implement in the case of `out`, but sometimes we have some other text. Would be good to have some common style.

File: odl/operator/default_ops.py [hunk 15]
```diff
@@ -579,6 +585,211 @@ def __str__(self):
...
+        1.0
+        """
+        point = self.domain.element(point)
+        diff = point - self.vector
+        dist = self.vector.dist(point)
+        if dist == 0:
+            raise ValueError('not differentiable at the reference vector {!r}'
+                             ''.format(self.vector))
+
+        return InnerProductOperator(diff / dist)
```
Comment (kohr-h) [inline odl/operator/default_ops.py]: Same here, need to exclude custom stuff.

**Comment by kohr-h** [comment]
Minor things to be fixed, otherwise alright.

File: odl/operator/default_ops.py (hunk 1 above)
Comment (adler-j) [inline odl/operator/default_ops.py]: > This assumes quite a bit on the norm of the space. The inner product operator needs to use the same weighting as the norm - I guess that's okay.

Sure, but this is a requirement on the space that we have written in the doc.

>  But a custom norm implementation cannot be handled consistently, that needs to be filtered out (and documented that it won't work).

Depends on the implementer I guess, here we can only assume the norm and inner product match.

File: odl/operator/default_ops.py (hunk 3 above)
Comment (adler-j) [inline odl/operator/default_ops.py]: It did in my online editor, but I'll change it

File: odl/operator/default_ops.py (hunk 2 above)
Comment (adler-j) [inline odl/operator/default_ops.py]: It felt good here, what else do we go for in the doc?

Comment (kohr-h) [inline odl/operator/default_ops.py]: We have `\partial`, e.g. in `landweber`. Feels a bit more robust against other stuff that may also want to use the letter `D`.

**Pull request merged**